lib/Clam/Properties/BndCheck.cc

```diff
@@ -0,0 +1,383 @@
+                   : "store")
+                         << ": check size and offset are in bounds: " << I
+                         << "\n");
+      var_t size = it->second.first;
+      var_t offset = it->second.second;
+      Type *elemTy = vptr->getType()->getPointerElementType();
+      const llvm::DataLayout *dl = getInsDatalayout(I);
+      unsigned size_of = dl->getTypeSizeInBits(elemTy) / 8;
+      bb.assertion(offset >= number_t(0), getDebugLoc(&I, m_assertionId++));
+      bb.assertion(size >= offset + size_of, getDebugLoc(&I, m_assertionId++));
+      m_addedChecks++;
+    }
+  }
+
+public:
+  EmitBndChecksImpl(const CrabBuilderParams &params, crabLitFactory &lfac,
+                    uint32_t &assertionId)
+      : m_params(params), m_lfac(lfac), m_assertionId(assertionId),
+        m_addedChecks(0), m_dl(nullptr) {}
+
+  ~EmitBndChecksImpl() {
+    llvm::errs() << "-- Inserted " << m_addedChecks;
+    llvm::errs() << " bound checks\n";
+    llvm::errs() << " (skipped " << m_trivialChecks << " trivial checks).\n";
+  }
+
+  void visitBeforeBasicBlock(llvm::BasicBlock &BB) {
+    // We cache checks only in the same basic block
+    m_seen.clear();
+  }
+
+  void visitBeforeStore(llvm::StoreInst &I, CrabStoreRefOps &s) {
+    if (!isRelevantCheck(I.getOperand(1))) {
+      m_trivialChecks++;
+    } else {
+      addBndCheckAssertion(I, s.getRegion(), s.getBasicBlock(), false);
+    }
+  }
+
+  void visitBeforeLoad(llvm::LoadInst &I, CrabLoadRefOps &s) {
+    if (!isRelevantCheck(I.getOperand(0))) {
+      m_trivialChecks++;
+    } else {
+      addBndCheckAssertion(I, s.getRegion(), s.getBasicBlock(), true);
+    }
+  }
+
+  void visitAfterLoad(llvm::LoadInst &I, CrabLoadRefOps &s) {
+    crab_lit_ref_t lhs = m_lfac.getLit(I);
+    basic_block_t &bb = s.getBasicBlock();
+    if (lhs->getVar().get_type().is_reference()) {
+      var_t ref_obj_sz = m_lfac.mkIntVar(getMaxWidthInBits(I));
+      // obtain an overappriximation of deref
+      // using havoc to initialize size and offset
+      bb.havoc(ref_obj_sz, "approx size when loading a ptr of ptr");
+      var_t ref_offset = m_lfac.mkIntVar(getMaxWidthInBits(I));
+      bb.havoc(ref_offset, "approx offset when loading a ptr of ptr");
+
+      // add pair to the map by llvm pointer
+      Value *ref_v = llvm::cast<llvm::Value>(&I);
+      m_ref_bnd_map.insert({ref_v, {ref_obj_sz, ref_offset}});
+    }
+  }
+
+  // -- alloc site ptr bound utilities construction
+  // From LLVM IR to crab IR we perform the following transformation:
+  // p := alloc(sz) => p := make_ref(region, sz)
+  // We use the idea from fat pointer like by adding two additional variables:
+  // p_offset := 0;
+  // p_size := sz;
+  // where offset indicates the offset of current pointer
+  // the size means the total size of allcation
+  void visitAfterAlloc(llvm::Instruction &I, const llvm::TargetLibraryInfo &tli,
+                       CrabMakeRefOps &s) {
+    // llvm::errs() << I << "\n";
```

Comment:
This code is complicated because you are trying to get the size of the allocated memory using `I`.
However, `s.getSize()` gives you already that!